ATB.BL/Services/BookingService.cs

```diff
@@ -0,0 +1,83 @@
+﻿using ATB.DataAccess;
+using ATB.Entities;
+namespace ATB.Services
+{
+    public class BookingService
+    {
+        private readonly IBookingRepository _bookingRepository;
+
+        public BookingService(IBookingRepository bookingRepository)
+        {
+            _bookingRepository = bookingRepository;
+        }
+
+        public IEnumerable<Booking> GetAllBookings()
+        {
+            return _bookingRepository.GetAllBookings();
+        }
+        public IEnumerable<Booking> FilterBookings(BookingSearchCriteria bookingSearchCriteria)
+        {
+            var allBookings = GetAllBookings();
+            return allBookings
+                .Where(booking =>
+                     (booking.Flight.Price.Equals(bookingSearchCriteria.Price) || (bookingSearchCriteria.Price is null)) &&
+                     (booking.Flight.DepartureDate.Equals(bookingSearchCriteria.DepartureDate) || (bookingSearchCriteria.DepartureDate is null)) &&
+                     (booking.Flight.FClass.Equals(bookingSearchCriteria.FClass) || (bookingSearchCriteria.FClass is null)) &&
+                     (booking.Flight.DestinationCountry.Equals(bookingSearchCriteria.DestinationCountry) || (bookingSearchCriteria.DestinationCountry is null)) &&
+                     (booking.Flight.ArrivalAirport.Equals(bookingSearchCriteria.ArrivalAirport) || (bookingSearchCriteria.ArrivalAirport is null)) &&
+                     (booking.Flight.DepartureCountry.Equals(bookingSearchCriteria.DepartureCountry) || (bookingSearchCriteria.DepartureCountry is null)) &&
+                     (booking.Flight.DepartureAirport.Equals(bookingSearchCriteria.DepartureAirport) || (bookingSearchCriteria.DepartureAirport is null)) &&
+
+                     (booking.Flight.FlightId.Equals(bookingSearchCriteria.FlightId) || (bookingSearchCriteria.FlightId is null)) &&
+                     (booking.Passenger.PassengerId.Equals(bookingSearchCriteria.PassengerId) || (bookingSearchCriteria.PassengerId is null))
+                     );
```

Comment:
this looks a bit messy, how can I make a bit more clean?

Reply:
refactored code #1 

 public IEnumerable<Booking> FilterBookings(BookingSearchCriteria bookingSearchCriteria)
        {
            var allBookings = GetAllBookings();

            bool IsNotSpecified<T>(T value) => value == null;

            bool FilterPredicate(Booking booking)
            {
                var flight = booking.Flight;

                bool IsPriceMatch = flight.Price.Equals(bookingSearchCriteria.Price) || IsNotSpecified(bookingSearchCriteria.Price);
                bool IsDepartureDateMatch = flight.DepartureDate.Equals(bookingSearchCriteria.DepartureDate) || IsNotSpecified(bookingSearchCriteria.DepartureDate);
                bool IsFClassMatch = flight.FClass.Equals(bookingSearchCriteria.FClass) || IsNotSpecified(bookingSearchCriteria.FClass);
                bool IsDestinationCountryMatch = flight.DestinationCountry.Equals(bookingSearchCriteria.DestinationCountry) || IsNotSpecified(bookingSearchCriteria.DestinationCountry);
                bool IsArrivalAirportMatch = flight.ArrivalAirport.Equals(bookingSearchCriteria.ArrivalAirport) || IsNotSpecified(bookingSearchCriteria.ArrivalAirport);
                bool IsDepartureCountryMatch = flight.DepartureCountry.Equals(bookingSearchCriteria.DepartureCountry) || IsNotSpecified(bookingSearchCriteria.DepartureCountry);
                bool IsDepartureAirportMatch = flight.DepartureAirport.Equals(bookingSearchCriteria.DepartureAirport) || IsNotSpecified(bookingSearchCriteria.DepartureAirport);
                bool IsFlightIdMatch = flight.FlightId.Equals(bookingSearchCriteria.FlightId) || IsNotSpecified(bookingSearchCriteria.FlightId);
                bool IsPassengerIdMatch = booking.Passenger.PassengerId.Equals(bookingSearchCriteria.PassengerId) || IsNotSpecified(bookingSearchCriteria.PassengerId);

                return IsPriceMatch && IsDepartureDateMatch && IsFClassMatch &&
                       IsDestinationCountryMatch && IsArrivalAirportMatch &&
                       IsDepartureCountryMatch && IsDepartureAirportMatch &&
                       IsFlightIdMatch && IsPassengerIdMatch;
            }

            var filteredBookings = allBookings.Where(FilterPredicate);

            return filteredBookings;
        }